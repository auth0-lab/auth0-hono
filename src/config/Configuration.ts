import { OIDCAuthorizationRequestParams } from "@/config/authRequest.js";
import { SessionConfiguration } from "@/types/session.js";
import { SessionOptions } from "@jfromaniello/hono-sessions";

type Routes = {
  login: string;
  logout: string;
  backchannelLogout?: string;
  callback: string;
};

export interface Configuration {
  /**
   * The base URL of the OIDC provider.
   */
  domain: string;

  /**
   * The base URL of the application.
   */
  baseURL: string;

  /**
   * The client ID of the application.
   */
  clientID: string;

  /**
   * The client secret of the application.
   */
  clientSecret?: string;

  /**
   * Whether to require authentication for all routes.
   * @default true
   */
  authRequired: boolean;

  /**
   * Whether to use the IDP's logout endpoint.
   * @default false
   */
  idpLogout: boolean;

  /**
   * hono-sessions options.
   *
   * @default {
   *  store: new CookieStore(),
   *  sessionCookieName: 'appSession',
   *  cookieOptions: {
   *    sameSite: 'Lax',
   *    path: '/',
   *    httpOnly: true,
   *  }
   * }
   */
  session: SessionConfiguration;

  /**
   * Use this setting to prevent the default routes from being installed.
   *
   * @default []
   */
  customRoutes: (keyof Routes)[];

  /**
   * Whether to mount the default routes.
   *
   * If set to false, you must manually define routes and use the middlewares
   * login(), callback(), and logout() in your application.
   *
   * You can disable individual routes by using the `customRoutes` setting.
   */
  mountRoutes: boolean;

  /**
   * Routes options.
   *
   * @default {
   *   login: '/login',
   *   logout: '/logout',
   *   callback: '/callback',
   * }
   */
  routes: Routes;

  /**
   * Additional authorization request parameters that will be included in
   * the authorization URL.
   *
   * @default {
   *  response_type: 'id_token',
   *  scope: 'openid profile email',
   *  response_mode: 'form_post',
   * }
   */
  authorizationParams: Partial<OIDCAuthorizationRequestParams>;

  /**
   * Forwards specific query parameters from the login request to the authorization request.
   * This allows passing through parameters like 'ui_locales', 'acr_values', or custom parameters
   * that your identity provider supports without having to specify them in authorizationParams.
   */
  forwardAuthorizationParams?: string[];

  /**
   * Additional parameters that will be sent to the
   * token endpoint, typically used for parameters such as `resource`
   * in cases where multiple resource indicators were requested but
   * the authorization server only supports issuing an access token
   * with a single audience
   */
  tokenEndpointParams?: Record<string, string>;

  /**
   * Whether to use pushed authorization requests.
   * @default false
   */
  pushedAuthorizationRequests: boolean;

  /**
   * The clock tolerance for the OIDC client.
   * @default 60
   */
  clockTolerance: number;

  /**
   * Whether to enable telemetry.
   * @default true
   */
  enableTelemetry: boolean;

  /**
   * The HTTP timeout for the OIDC client.
   * @default 5000
   */
  httpTimeout?: number;

  // Hooks
  // afterCallback?: (
  //   c: Context,
  //   session: OidcSession,
  //   state: any,
  // ) => Promise<OIDCUserInfoResponse> | OIDCUserInfoResponse;

  /**
   * The method to use for client authentication.
   *
   * If `authorizationParams.response_type` is `id_token` and !pushedAuthorizationRequests,
   * the @default is `none`.
   *
   * If `clientAssertionSigningKey` is provided,
   *  the @default is `private_key_jwt`.
   *
   * Otherwise, the @default is `client_secret_basic`.
   */
  clientAuthMethod:
    | "client_secret_basic"
    | "client_secret_post"
    | "client_secret_jwt"
    | "private_key_jwt"
    | "none";

  /**
   * The client assertion signing key.
   * Required if `clientAuthMethod` is `private_key_jwt`.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  clientAssertionSigningKey?: any;

  /**
   * The client assertion signing algorithm.
   * Required if `clientAuthMethod` is `private_key_jwt`.
   * @default 'RS256'
   */
  clientAssertionSigningAlg?:
    | "RS256"
    | "RS384"
    | "RS512"
    | "PS256"
    | "PS384"
    | "PS512"
    | "ES256"
    | "ES256K"
    | "ES384"
    | "ES512"
    | "EdDSA";

  /**
   * Returns 401 if the user is not authenticated.
   * @default false
   */
  errorOnRequiredAuth: boolean;

  /**
   * Whether to attempt a silent login.
   * @default false
   */
  attemptSilentLogin: boolean;

  /**
   * The claims to exclude from the user identity.
   * @default ['aud', 'iss', 'iat', 'exp', 'nbf', 'nonce', 'azp', 'auth_time', 's_hash', 'at_hash', 'c_hash']
   */
  excludedClaims: string[];

  /**
   * The expected signing algorithm for the ID token.
   * @default 'RS256'
   */
  idTokenSigningAlg: string;

  /**
   * The maximum age of the discovery cache.
   * @default 10 * 60 * 1000
   */
  discoveryCacheMaxAge: number;

  /**
   * The HTTP user agent to use for the OIDC client.
   * @default 'hono-openid-connect'
   */
  httpUserAgent: string;

  /**
   * The fetch function to use for the OIDC client.
   * @default globalThis.fetch
   */
  fetch: typeof globalThis.fetch;

  /**
   * Additional parameters that will be sent to the
   * logout endpoint.
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  logoutParams?: Record<string, any>;

  /**
   * Logger function to use for the OIDC client.
   * @param message - The message to log.
   * @param metadata - Additional metadata to include in the log.
   */
  debug: (message: string, metadata?: Record<string, unknown>) => void;
}

// Type for the required fields that must be provided in InitConfiguration
type RequiredConfigFields = "domain" | "baseURL" | "clientID";

// Type for the optional session field that should be partial in InitConfiguration
type SessionField = {
  session?: false | Partial<SessionOptions>;
};

// Type for the optional routes field that should be partial in InitConfiguration
type RoutesField = {
  routes?: Partial<Routes>;
};

/**
 * Configuration type for initializing the OIDC client.
 * This represents the input before validation, where most properties are optional
 * since they have defaults in the schema.
 */
export type InitConfiguration = Pick<Configuration, RequiredConfigFields> &
  Partial<Omit<Configuration, RequiredConfigFields | "session" | "routes">> &
  SessionField &
  RoutesField;
