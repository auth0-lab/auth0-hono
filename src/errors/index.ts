export class MissingContextError extends Error {
  public code: string = "missing_store_options_error";

  constructor(message?: string) {
    super(message ?? "The context options is missing.");
    this.name = "MissingContextError";
  }
}
