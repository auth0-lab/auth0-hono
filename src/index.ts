export { auth } from "@/auth.js";

export type { OIDCEnv, OIDCVariables } from "@/lib/honoEnv.js";
export { type UserInfoResponse as UserInfo } from "openid-client";

export {
  attemptSilentLogin,
  backchannelLogout,
  callback,
  login,
  logout,
  pauseSilentLogin,
  requiresAuth,
  resumeSilentLogin,
} from "@/middleware/index.js";

export type { TokenEndpointResponse as TokenSet } from "openid-client";

export { Auth0Exception } from "@/lib/Exception.js";
