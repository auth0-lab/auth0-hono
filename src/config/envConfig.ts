import { InitConfiguration } from "@/config/Configuration.js";
import { MakeOptional } from "@/types/util.js";

export type MinimalConfigByEnv = {
  AUTH0_DOMAIN: string;
  AUTH0_CLIENT_ID: string;
  AUTH0_CLIENT_SECRET?: string;
  AUTH0_AUDIENCE?: string;
  BASE_URL: string;
  AUTH0_SESSION_ENCRYPTION_KEY?: string;
};

export type PartialConfig = MakeOptional<
  InitConfiguration,
  "clientID" | "clientSecret" | "domain" | "baseURL" | "session"
>;

export const envHasConfig = (
  config: MinimalConfigByEnv | unknown,
): config is MinimalConfigByEnv => {
  return (
    typeof config === "object" &&
    config !== null &&
    "AUTH0_DOMAIN" in config &&
    "AUTH0_CLIENT_ID" in config &&
    "BASE_URL" in config &&
    typeof config.AUTH0_DOMAIN === "string" &&
    typeof config.AUTH0_CLIENT_ID === "string" &&
    typeof config.BASE_URL === "string"
  );
};

export const assignFromEnv = (
  config: PartialConfig,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  env: Record<string, any>,
): InitConfiguration => {
  const configWithoutEnv = config ?? ({} as PartialConfig);
  if (!envHasConfig(env)) {
    return configWithoutEnv as InitConfiguration;
  }

  const {
    AUTH0_DOMAIN,
    AUTH0_CLIENT_ID,
    AUTH0_CLIENT_SECRET,
    BASE_URL,
    AUTH0_AUDIENCE,
  } = env;

  const authorizationParams = {...configWithoutEnv.authorizationParams};
  authorizationParams.audience = authorizationParams.audience ?? AUTH0_AUDIENCE;

  return {
    ...configWithoutEnv,
    domain: configWithoutEnv.domain ?? AUTH0_DOMAIN,
    clientID: configWithoutEnv.clientID ?? AUTH0_CLIENT_ID,
    clientSecret: configWithoutEnv.clientSecret ?? AUTH0_CLIENT_SECRET,
    baseURL: configWithoutEnv.baseURL ?? BASE_URL,
    authorizationParams: authorizationParams,
    session: {
      ...(configWithoutEnv.session || {}),
      secret:
        configWithoutEnv.session?.secret ??
        process.env.AUTH0_SESSION_ENCRYPTION_KEY,
    },
  };
};
