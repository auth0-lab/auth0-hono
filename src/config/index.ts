import { OIDCEnv } from "@/lib/honoEnv.js";
import { Context } from "hono";
import { HTTPException } from "hono/http-exception";
import { Configuration, InitConfiguration } from "./Configuration.js";
import { ConfigurationSchema } from "./Schema.js";

const parsedConfig = new Map<InitConfiguration, Configuration>();

export const parseConfiguration = (
  config: InitConfiguration,
): Configuration => {
  if (parsedConfig.has(config)) {
    return parsedConfig.get(config)!;
  }
  const result = ConfigurationSchema.parse(config) as Configuration;
  parsedConfig.set(config, result);
  return result;
};
export {
  assignFromEnv,
  type ConditionalInitConfig,
} from "@/config/envConfig.js";

export const getClient = (c: Context<OIDCEnv>) => {
  if (!c.var.auth0Client || !c.var.auth0Configuration) {
    throw new HTTPException(500, {
      message:
        "The auth0 middleware is not properly configured. Install `auth` first.",
    });
  }
  return {
    client: c.var.auth0Client,
    configuration: c.var.auth0Configuration,
  };
};
