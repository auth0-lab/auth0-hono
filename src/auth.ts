import { assignFromEnv, parseConfiguration } from "@/config/index.js";
import { initializeOidcClient } from "@/lib/client.js";
import { OIDCEnv } from "@/lib/honoEnv.js";
import {
  backchannelLogout as backchannelLogoutHandler,
  callback as callbackHandler,
  login as loginHandler,
  logout as logoutHandler,
  requiresAuth,
} from "@/middleware/index.js";
import { ServerClient } from "@auth0/auth0-server-js";
import { Context, MiddlewareHandler, Next } from "hono";
import { every } from "hono/combine";
import { createMiddleware } from "hono/factory";
import { Configuration } from "./config/Configuration.js";
import { PartialConfig } from "./config/envConfig.js";
import { HonoCookieHandler } from "./session/HonoCookieHandler.js";

/**
 * Main auth middleware function.
 *
 * This function initializes the OIDC middleware with the provided configuration.
 * It sets up the session middleware if needed and handles the OIDC client initialization.
 * It also manages the routing for login, callback, and logout endpoints.
 *
 */
export function auth(initConfig: PartialConfig = {}): MiddlewareHandler {
  let client: ServerClient<Context>;
  let config: Configuration;
  // Main OIDC middleware function
  const oidcMiddleware: MiddlewareHandler = createMiddleware<OIDCEnv>(
    async (c, next: Next): Promise<Response | void> => {
      try {
        if (!client) {
          // Initialize the client
          const withEnvVars = assignFromEnv(initConfig, {
            ...c.env,
            ...(process.env ?? {}),
          });
          config = parseConfiguration(withEnvVars);
          client = initializeOidcClient(config);
        }
        c.set("auth0Client", client);
        c.set("auth0Configuration", config);

        // Use destructuring with defaults to ensure routes is always defined
        const { routes, authRequired, mountRoutes } = config;
        const { login, callback, logout, backchannelLogout } = routes;

        // Handle login route
        if (
          mountRoutes &&
          !config.customRoutes.includes("login") &&
          c.req.path === login &&
          c.req.method === "GET"
        ) {
          return loginHandler()(c, next);
        }

        // Handle callback route
        if (
          mountRoutes &&
          !config.customRoutes.includes("callback") &&
          c.req.path === callback
        ) {
          return callbackHandler()(c, next);
        }

        // Handle logout route
        if (
          mountRoutes &&
          !config.customRoutes.includes("logout") &&
          c.req.path === logout &&
          c.req.method === "GET"
        ) {
          return logoutHandler()(c, next);
        }

        // Handle backchannel logout route
        if (
          mountRoutes &&
          !config.customRoutes.includes("backchannelLogout") &&
          c.req.path === backchannelLogout &&
          c.req.method === "POST"
        ) {
          return backchannelLogoutHandler()(c, next);
        }

        // Handle unauthenticated requests
        if (authRequired) {
          return requiresAuth()(c, next);
        }
      } catch (error) {
        console.error("AUTH0 Middleware Error:", error);
        return c.text("Internal Server Error", 500);
      }
      // // Continue to the next middleware or route handler
      return await next();
    },
  );

  const setHonoContext = createMiddleware(async (c, next) => {
    return HonoCookieHandler.setContext(c, next);
  });

  return every(setHonoContext, oidcMiddleware);
}
