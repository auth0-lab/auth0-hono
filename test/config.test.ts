/* eslint-disable @typescript-eslint/no-explicit-any */
import { describe, expect, it } from "vitest";
import { parseConfiguration } from "../src/config";
import { InitConfiguration } from "../src/config/Configuration";
import { assignFromEnv } from "../src/config/envConfig";

describe("Configuration Parser", () => {
  it("should parse a valid configuration", () => {
    const validConfig: InitConfiguration = {
      domain: "auth.example.com",
      baseURL: "https://app.example.com",
      clientID: "test-client-id",
      clientSecret: "test",
      session: {
        secret: "test encryption key fdsgfds gfds ",
      },
    };

    const parsedConfig = parseConfiguration(validConfig);

    expect(parsedConfig).toHaveProperty("domain", "auth.example.com");
    expect(parsedConfig).toHaveProperty("baseURL", "https://app.example.com");
    expect(parsedConfig).toHaveProperty("clientID", "test-client-id");
    expect(parsedConfig).toHaveProperty("authRequired", true); // default value
    expect(parsedConfig).toHaveProperty("fetch", globalThis.fetch); // default value
    expect(parsedConfig).toMatchSnapshot();
  });

  it("should throw an error for invalid configuration", () => {
    const invalidConfig = {
      // Missing required domain
      baseURL: "https://app.example.com",
      clientID: "test-client-id",
    };

    expect(() => parseConfiguration(invalidConfig as any)).toThrow();
  });

  it("should apply default values to configuration", () => {
    const minimalConfig: InitConfiguration = {
      domain: "auth.example.com",
      baseURL: "https://app.example.com",
      clientID: "test-client-id",
      clientSecret: "test-client-secret",
      session: {
        secret: "test encryption key fdsgfds gfds ",
      },
    };

    const parsedConfig = parseConfiguration(minimalConfig);

    expect(parsedConfig).toHaveProperty("authRequired", true);
    expect(parsedConfig).toHaveProperty("clockTolerance", 60);
    expect(parsedConfig).toHaveProperty("routes");

    expect(parsedConfig.routes).toEqual({
      login: "/auth/login",
      logout: "/auth/logout",
      callback: "/auth/callback",
      backchannelLogout: "/auth/backchannel-logout",
    });
  });

  it("should cache parsed configurations", () => {
    const config: InitConfiguration = {
      domain: "auth.example.com",
      baseURL: "https://app.example.com",
      clientID: "test-client-id",
      clientSecret: "test",
      session: {
        secret: "test encryption key fdsgfds gfds ",
      },
    };

    const firstParsed = parseConfiguration(config);
    const secondParsed = parseConfiguration(config);

    expect(firstParsed).toBe(secondParsed); // Should be the same object instance (cached)
  });

  it("should not allow custom routes to be set to relative paths", () => {
    const config: InitConfiguration = {
      domain: "auth.example.com",
      baseURL: "https://app.example.com",
      clientID: "test-client-id",
      routes: {
        login: "login",
      },
      session: {
        secret: "test encryption key fdsgfds gfds ",
      },
    };

    expect(() => parseConfiguration(config)).toThrow();
  });
});

describe("Environment Configuration", () => {
  describe("assignFromEnv", () => {
    const validEnv = {
      AUTH0_DOMAIN: "test.auth0.com",
      AUTH0_CLIENT_ID: "test-client-id",
      BASE_URL: "https://example.com",
      AUTH0_CLIENT_SECRET: "test-secret",
      AUTH0_AUDIENCE: "https://api.example.com",
    };

    it("should assign audience from env to authorizationParams when audience only exists in env and not config", () => {
      const config = {};
      const env = validEnv;

      const result = assignFromEnv(config, env);

      expect(result.authorizationParams).toEqual({
        audience: "https://api.example.com",
      });
    });

    it("should prioritize config authorizationParams.audience over env audience when both exist", () => {
      const config = {
        authorizationParams: {
          audience: "https://config-audience.com",
          scope: "openid profile",
        },
      };
      const env = validEnv;

      const result = assignFromEnv(config, env);

      expect(result.authorizationParams).toEqual({
        audience: "https://config-audience.com",
        scope: "openid profile",
      });
    });

    it("should propagate all other authorizationParams from config and merge with env audience if there is no config audience", () => {
      const config = {
        authorizationParams: {
          scope: "openid profile email",
          prompt: "login",
          max_age: 3600,
          ui_locales: "en-US",
        },
      };
      const env = validEnv;

      const result = assignFromEnv(config, env);

      expect(result.authorizationParams).toEqual({
        scope: "openid profile email",
        prompt: "login",
        max_age: 3600,
        ui_locales: "en-US",
        audience: "https://api.example.com",
      });
    });

    it("should assign env audience when config has empty authorizationParams object", () => {
      const config = {
        authorizationParams: {},
      };
      const env = validEnv;

      const result = assignFromEnv(config, env);

      expect(result.authorizationParams).toEqual({
        audience: "https://api.example.com",
      });
    });

    it("should assign env audience when config has no authorizationParams defined", () => {
      const config = {};
      const env = validEnv;

      const result = assignFromEnv(config, env);

      expect(result.authorizationParams).toEqual({
        audience: "https://api.example.com",
      });
    });

    it("should not set audience when audience is absent from both env and config", () => {
      const config = {
        authorizationParams: {
          scope: "openid profile",
        },
      };
      const envWithoutAudience = {
        AUTH0_DOMAIN: "test.auth0.com",
        AUTH0_CLIENT_ID: "test-client-id",
        BASE_URL: "https://example.com",
      };

      const result = assignFromEnv(config, envWithoutAudience);

      expect(result.authorizationParams).toEqual({
        scope: "openid profile",
      });
    });

    it("should handle undefined config gracefully and use env values", () => {
      const config = undefined;
      const env = validEnv;

      const result = assignFromEnv(config, env);

      expect(result.authorizationParams).toEqual({
        audience: "https://api.example.com",
      });
      expect(result.domain).toBe("test.auth0.com");
      expect(result.clientID).toBe("test-client-id");
      expect(result.baseURL).toBe("https://example.com");
    });

    it("should return config as-is when env doesn't have required fields", () => {
      const config = {
        domain: "config.auth0.com",
        clientID: "config-client-id",
        authorizationParams: {
          scope: "openid profile",
        },
      };
      const invalidEnv = {
        // Missing required AUTH0_DOMAIN, AUTH0_CLIENT_ID, BASE_URL
        AUTH0_AUDIENCE: "https://api.example.com",
      };

      const result = assignFromEnv(config, invalidEnv);

      expect(result).toEqual(config);
    });
  });
});
