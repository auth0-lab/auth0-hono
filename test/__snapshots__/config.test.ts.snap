// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Configuration Parser > should parse a valid configuration 1`] = `
{
  "attemptSilentLogin": false,
  "authRequired": true,
  "authorizationParams": {
    "response_type": "code",
    "scope": "openid profile email",
  },
  "baseURL": "https://app.example.com",
  "clientAuthMethod": "client_secret_basic",
  "clientID": "test-client-id",
  "clientSecret": "test",
  "clockTolerance": 60,
  "customRoutes": [],
  "debug": [Function],
  "discoveryCacheMaxAge": 600000,
  "domain": "auth.example.com",
  "enableTelemetry": true,
  "errorOnRequiredAuth": false,
  "excludedClaims": [
    "aud",
    "iss",
    "iat",
    "exp",
    "nbf",
    "nonce",
    "azp",
    "auth_time",
    "s_hash",
    "at_hash",
    "c_hash",
  ],
  "fetch": [Function],
  "forwardAuthorizationParams": [],
  "httpTimeout": 5000,
  "httpUserAgent": "hono-openid-connect",
  "idTokenSigningAlg": "RS256",
  "idpLogout": false,
  "mountRoutes": true,
  "pushedAuthorizationRequests": false,
  "routes": {
    "backchannelLogout": "/auth/backchannel-logout",
    "callback": "/auth/callback",
    "login": "/auth/login",
    "logout": "/auth/logout",
  },
  "session": {
    "absoluteDuration": 259200,
    "cookie": {
      "name": "appSession",
      "sameSite": "lax",
      "secure": true,
    },
    "inactivityDuration": 86400,
    "rolling": true,
    "secret": "test encryption key fdsgfds gfds ",
  },
}
`;
